### DISCUSSION TITLE
2D ERT Inversion in cone

### DISCUSSION THREAD

**Discussion by bralec** [gimli-org/pyGIMLi#330] [Q&A] [answered]
Cordial greeting.

I am trying to visualize the 2D ERT inversion in cone, as shown in the example of BERT: http://resistivity.net/bert/_examples/plot_2D_simple.html#sphx-glr-examples-plot-2d-simple-py

In that example of BERT, the mesh leaves blank spaces, but when I run this example of BERT in pyGIMLi, the white spaces fill up, and I need you not to. I want it to look like the example of BERT, like a 'cone'.

If you can help me, I’d appreciate it very much.

**Comment by halbmy** [comment, marked as answer]
Sorry for the late response due to vacations, field surveys and conferences. Actually, the documentation on resistivity.net is totally outdated. For standard ERT you don't need pybert anymore but the ert module from pyGIMLi. I've made a few modifications to the script (1. replacing pb by ert and 2. some stuff concerning the grid cell numeration and markering):
```
# -*- coding: utf-8 -*-
from pygimli.physics import ert
import pygimli as pg
import pygimli.meshtools as mt  # save space

world = mt.createWorld(start=[-50, 0], end=[50, -50], layers=[-1, -5],
                       worldMarker=True)

# Create some heterogeneous circular anomaly
block = mt.createCircle(pos=[0, -3.], radius=1, marker=4, boundaryMarker=10,
                        area=0.1)

# Merge geometry definition into a Piecewise Linear Complex (PLC)
geom = mt.mergePLC([world, block])

# Optional: show the geometry
pg.show(geom)

scheme = ert.createData(elecs=pg.utils.grange(start=-10, end=10, n=21),
                       schemeName='dd')

for pos in scheme.sensorPositions():
    geom.createNode(pos)
    geom.createNode(pos + pg.RVector3(0, -0.1))

mesh = mt.createMesh(geom, quality=34)

# Create a map to set resistivity values in the appropriate regions
# [[regionNumber, resistivity], [regionNumber, resistivity], [...]
rhomap = [[1, 100.],
          [2, 50.],
          [3, 10.],
          [4, 100.]]

# Optional: take a look at the mesh
pg.show(mesh, data=rhomap, label='Resistivity $(\Omega$m)', showMesh=True)

mgr = ert.ERTManager()

# Perform the modeling with the mesh and the measuring scheme itself
# and return a data container with apparent resistivity values,
# geometric factors and estimated data errors specified by the noise setting.
# The noise is also added to the data.
data = ert.simulate(mesh, res=rhomap, scheme=scheme, noiseLevel=1,
                    noiseAbs=1e-6)

# Optional: you can filter all values and tokens in the data container.
print('Simulated rhoa', data('rhoa'), max(data('rhoa')))

# Its possible that there are some negative data values due to noise and
# huge geometric factors. So we need to remove them
data.markInvalid(data('rhoa') < 0)
print('Filtered rhoa', data('rhoa'), max(data('rhoa')))
data.removeInvalid()

# Optional: save the data for further use
data.save('simple.dat')

# Optional: take a look at the data
ert.show(data)

# Run the ERTManager to invert the modeled data.
# The necessary inversion mesh is generated automatic.
model = mgr.invert(data, paraDX=0.3, maxCellArea=0.2, lam=20)

# Let the ERTManger show you the model and fitting results of the last
# successful run.
# Show data, model response, and model
mgr.showResultAndFit()
# %%
import numpy as np
x = pg.utils.grange(start=-12, end=12, n=33)
# y = -pg.utils.grange(0.5, 8, n=8, log=True)
# y = pg.cat(pg.Vector(1, 0.), y)
y = np.arange(-8., 0.1, 1.0)
print(y)
grid = pg.createGrid(x=x, y=y, marker=1)
# %
mesh = pg.meshtools.appendTriangleBoundary(grid, xbound=50, ybound=50, marker=0)
pg.show(mesh, markers=True)
# %%
model = mgr.invert(data, mesh=mesh, lam=20)
mgr.showResultAndFit()

chi2 = mgr.inv.getChi2()

print(chi2)
# Stop the script here and wait until all figure are closed.
pg.wait()
```
Then all seems to work well.
![image](https://user-images.githubusercontent.com/7136401/132391076-1488610f-5053-4a0f-9b9d-78c1e27b1e12.png)
![image](https://user-images.githubusercontent.com/7136401/132391093-9bf7dab8-21f5-4acb-962b-c334af798d5d.png)

**Reply by bralec** [reply]
Don't worry, on the contrary, I really appreciate the answer.

Yes, it is true, as you indicate everything seems to work well, thank you very much for showing me this transition to the current inversion of pyGIMLi and how to model the previous version (pyBERT).

However, I get a doubt. The picture of the inversion shows two-zone without information (I put it with red lines because they are white zones); Is it possible to do that too? I mean, decide when I want that blank information filled since I see that inversions always do that filling (something like an interpolation, I think)

Thanks for everything.

PD: I find it very interesting as the transition significantly improved the result of chi2 (practically 1); it gives more security.

![Sin título](https://user-images.githubusercontent.com/82900508/132418877-8f1f9e7d-f04a-4e67-b247-6d39ab6d4c9f.png)

**Comment by halbmy** [comment]
Well, inversion finds a resistivity for all predefined model cells by using the regularization term (usually smoothness), so it will be kind of interpolated in regions of less resolution.

When using irregular triangle meshes, on could try to make the (inversion) model domain smaller (or cells bigger) to avoid badly resolved cells but then you could end up in not fitting your data as the structures causing your data anomalies are outside of the domain (or cells being too large). At any rate, resistivity is extrapolated outside of the inversion domain for the forward calculation with some assumption. So in my feeling it is better to have some freedom in the model and to cut badly resolved cells instead of running into other problems. Of course you can also use a binary threshold by `mgr.showResult(..., coverage=mgr.coverage()>-0.5)`
![image](https://user-images.githubusercontent.com/7136401/132459724-c0cfc0a6-bbff-4605-9da2-b55cfc912b8c.png)

**Comment by bralec** [comment]
Excellent, I already understand much better how it is interpreted and it turns out this whole system. It is super interesting.

Thank you so much for your help, Thomas.
